keps/sig-architecture/20191001-prevent-permabeta.md

```diff
@@ -0,0 +1,140 @@
+where users and vendors start building important infrastructure against APIs that are not considered stable.
+In addition, once a feature is on by default, the incentive to further stabilize appears to diminish.
+See the features that have been beta for a long time: CSRs and Ingresses as examples.
+If we're honest with ourselves, a single actor has been cleaning up behind a lot of the project to unstick perma-beta APIs.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+
+### Goals
+
+1. Prevent APIs and features from being in a single beta version for more than six months.
+2. Prevent beta APIs from being treated as GA by users and vendors.
+
+### Non-Goals
+
+1. Promote APIs to GA before they are ready.
+
+## Proposal
+
+Once an API or feature reaches beta, it has six months to 
+1. reach GA and deprecate the beta or 
+2. have a new beta version and deprecate the previous beta.
+
+If neither of those conditions met, the beta API/feature is deprecated in the second release with a stated intent to remove the feature entirely.
+To avoid removal, the feature must create a new beta version (it cannot go directly from deprecated to GA).
+
+For example, in v1.16, v1beta1 is released. Sample release note and API doc:
+> * "The v1beta1 version of this API will be evaluated during v1.16 and v1.17, then deprecated in v1.18 (in favor of a new beta version, a GA version, or with no replacement), then removed in v1.21"
+
+Scenario A - progression to v1beta2 in v1.18. Sample release note and API doc:
+> * "The v1beta1 version of this API is deprecated in favor of v1beta2, and will be removed in v1.21"
+> * "The v1beta2 version of this API will be evaluated during v1.18 and v1.19, then deprecated in v1.20 (in favor of a new beta version, a GA version, or with no replacement), then removed in v1.23"
+
+Scenario B - progression to v1 in v1.18. Sample release note and API doc:
+> * "The v1beta1 version of this API is deprecated in favor of v1, and will be removed in v1.21"
+
+Scenario C - deprecation with no replacement. Sample release note and API doc:
+> * "The v1beta1 version of this API is deprecated with no replacement, and will be removed in v1.21"
+
+By regularly having new beta versions, we can ensure that consumers will not grow long running dependencies on particular betas which could pin design decisions.
+It will also create an incentive for feature authors to push their features to GA instead of letting them live in a permanent beta state. 
+
+### Risks and Mitigations
+
+What are the risks of this proposal and how do we mitigate.
+Think broadly.
+For example, consider both security and how this will impact the larger kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+
+## Drawbacks [optional]
+
+Why should this KEP _not_ be implemented.
```

Comment:
Imposing any cost to keeping an API or feature in beta will provide an incentive to promote the API/feature to GA prematurely, even if it has significant gaps, such as the CSR API does.

Reply:
I agree that could be incentivized, which is why the production readiness work is vital

Reply:
Along those lines we will be kicking off that effort ASAP, hoping to meet on it before Wednesday next week. The discussion would include whether there is enough of this sort of effort to make a new sub project in sig-arch (or not). Would love to have interested parties involved...see https://groups.google.com/d/msg/kubernetes-sig-architecture/RWSzgT3S4MQ/uSEs9UFzCAAJ